server/api/auth.py

```diff
@@ -0,0 +1,150 @@
+import functools
+import json
+from sqlalchemy.orm.exc import NoResultFound
+from flask import Blueprint as bp
+from flask import (
+    make_response, redirect,
+    request, session,
+    url_for, g
+)
+from flask_jwt_extended import (
+    JWTManager, jwt_required, create_access_token,
+    jwt_refresh_token_required, create_refresh_token,
+    get_jwt_identity, set_access_cookies,
+    set_refresh_cookies, unset_jwt_cookies,
+    jwt_optional, verify_jwt_in_request
+)
+from models.user import User
+
+auth_views = bp('auth', __name__, url_prefix="/auth")
+
+def authenticate(view):
+    @functools.wraps(view)
+    def decorated_function(*args, **kwargs):
+        # First look for a user object
+        if g.user is None:
+            return {
+                "status": 401,
+                "message": "You have to be logged in for that"
+            }
+        # Then validate the token
+        try:
+            # This will raise an error if the token
+            # is not valid is some way
+            verify_jwt_in_request()
+        except:
+            return {
+                'status': 401,
+                'message': "Invalid Token"
+            }
+        return view(*args, **kwargs)
+    return decorated_function
+
+@auth_views.before_app_request
+def load_logged_in_user():
+    user_id = session.get("user_id", None)
+    if user_id is not None:
+        g.user = User.get_instance(**{'id': user_id})
+    else:
+        g.user = None
+
+@auth_views.route("/signup", methods=["POST"])
+def signup():
+    if request.method == "POST":
+        request_dict = request.form.to_dict()
+        user_info = {}
+        for field in User.get_fields():
+            if field == "id":
+                continue
+            if field == "password_hash":
+                user_info['password'] = request_dict['password']
+                continue
+            user_info[field] = request_dict[field]
+        try:
+            user = User.create(**user_info)
```

Comment:
this will automatically let us know if the user already exists right?

Reply:
Yes it will, it has a method in it to check if a user with the given username or email exists already.